### ISSUE TITLE
enable non-ivy/maven repos

### ISSUE THREAD

**Issue by hochgi** [coursier/coursier#600]
(originally discussed over [gitter](https://gitter.im/coursier/coursier?at=59533c41ceb5bef82ebaf760))  

It would be great if coursier can be used on custom repositories and custom artifacts.
for instance, Apache dist repository, mirrored at either http://www-eu.apache.org/dist or http://www-us.apache.org/dist/ hosts various artifacts distributions. Sometimes, such artifacts are needed to be fetched as dependencies in SBT. 

e.g: let's take zookeeper as an example.
assuming I want version 3.4.9 tar.gz artifact, which is hosted here:  
http://www-us.apache.org/dist/zookeeper/zookeeper-3.4.9/  

theres 4 files there, the tar.gz, md5, sha1, asc.
No pom or ivy though... but, if it would be possible to define a custom resolver, something like the following pseudo code:
```scala
customCoursierResolvers += CustomCoursierResolver.url("http://www-us.apache.org/dist")
                                                 .pattern("/[organisation]/[module]-[revision]/[artifact].[ext]")
                                                 .validate( /* Some way to ensure checksum and signature validation... */ )
```
which will then enable to just add such custom dependency:
```scala
libraryDependencies += zookeeper" % "zookeeper" % "3.4.9" artifacts (Artifact("zookeeper", "tar.gz", "tar.gz")) intransitive()

classpathTypes ~=  (_ + "tar.gz")
```
maybe even enable multiple patterns, since in this case, same dist repository hosts other projects with different pattern. e.g. kafka ( http://www-us.apache.org/dist/kafka/0.11.0.0/ ):
```text
kafka pattern = "/[organisation]/[revision]/[artifact].[ext]"
```
it also hosts more artifacts (sources, docs, etc'...)

bottom line, a way to customize how coursier is interacting with various repositories would be great to have.

**Comment by alexarchambault** [comment]
The problem with these repositories is that they don't seem to have metadata files... Note that you could use the coursier cache API directly, if you're just interested in downloading / caching / verifying checksums (via [coursier.Cache.file](https://github.com/coursier/coursier/blob/dd3e8be410fc908790a036ac136d1946ef1f8ad1/cache/src/main/scala/coursier/Cache.scala#L941) in particular).

**Comment by hochgi** [comment]
Wasn't aware I could use the cache directly. So thanks, it's useful, and will probably use it.

Regarding my original feature request, I know some meta data is missing, but thought maybe it could be provided by user with sensible defaults? 
e.g: dependencies could be defined as a `Seq[ModuleID]` and defaults to `Nil`\*.

Also, verification by checking checksum match, could be taken from repo. (I'm assuming cousier checks downloaded files against hash & signature files)

\* maybe a better type would be so.ething like: `() => Future[Seq[ModuleID]]` to enable fetching dependencies dynamically.

**Comment by alexarchambault** [comment]
Closing this for now. Don't hesitate to send a PR with that kind of feature, or to discuss that further here.

**Comment by hochgi** [comment]
@alexarchambault thanks. I ended up using coursier cache as you suggested.
That code lies here:
https://github.com/CM-Well/CM-Well/blob/master/server/project/cmwell-build-plugin/src/main/scala/cmwell/build/CMWellBuild.scala

And it was very useful.
Thanks for the advice, and I guess future searchers getting to this issue can benefit from that advice as well.
